File: DataStructures/LinkedList.js
"use strict";

class LinkedList1 {

    constructor() {

        this.root = null;
        this.length = 0;

    }

    add(item) {

        if (!this.root) {

            this.root = {
                data: item,
                nextNode: null
            };

        } else {

            let node = this.root;

            while (node.nextNode) {

                node = node.nextNode;

            }

            node.nextNode = {
                data: item,
                nextNode: null
            };

        }

        this.length++;

    }

    addFront(item) {

        this.root = {
            data: item,
            nextNode: this.root
        };

        this.length++;

    }

    remove() {

        let node = this.root;

        if (node.nextNode) {

            while (node.nextNode.nextNode) {

                node = node.nextNode;

            }

            node.nextNode = null;


        } else {

            this.root = null;

        }

        this.length--;

    }

    removeFront() {

        this.root = this.root.nextNode;

        this.length--;

    }


    get(index) {

        let node = this.root;

        for (let i = 0; i < index; i++) {

            node = node.nextNode;

        }

        return node.data;

    }

    set(index, item) {

        let node = this.root;

        for (let i = 0; i < index; i++) {

            node = node.nextNode;

        }

        node.data = item;

    }

}

for (let linkedListClass of [LinkedList1]) {

    let linkedList = new linkedListClass();

    console.log(linkedList.length);

    linkedList.add(1);
    linkedList.add(2);
    linkedList.add(3);

    console.log(linkedList.get(2));

    linkedList.set(1, 4);

    console.log(linkedList.get(1));

    console.log(linkedList.get(0));

    linkedList.removeFront();

    linkedList.add(5);

    linkedList.remove();

    console.log(linkedList.get(0));

    linkedList.addFront(6);

    console.log(linkedList.get(0));

    console.log(linkedList.get(1));

    linkedList.add(7);
    linkedList.add(8);

    console.log(linkedList.length);

    console.log("------");

}

File: DataStructures/LinkedList.py
from collections import deque

class Node(object):

    def __init__(self, data, next_node=None):

        self.data = data
        self.next_node = next_node

class LinkedList1(object):

    def __init__(self):

        self.root = None
        self.length = 0

    def add(self, item):

        if not self.root:

            self.root = Node(item)

        else:

            node = self.root

            while node.next_node:

                node = node.next_node

            node.next_node = Node(item)

        self.length += 1

    def add_front(self, item):

        self.root = Node(item, self.root)

        self.length += 1

    def remove(self):

        node = self.root

        if node.next_node:

            while node.next_node.next_node:

                node = node.next_node

            node.next_node = None

        else:

            self.root = None

        self.length -= 1

    def remove_front(self):

        self.root = self.root.next_node

        self.length -= 1

    def __getitem__(self, index):

        node = self.root

        for i in range(index):

            node = node.next_node

        return node.data

    def __setitem__(self, index, item):

        node = self.root

        for i in range(index):

            node = node.next_node

        node.data = item

    def __len__(self):

        return self.length



class LinkedList2(object):

    def __init__(self):

        self.linkedlist = deque()

    def add(self, item):

        self.linkedlist.append(item)

    def add_front(self, item):

        self.linkedlist.appendleft(item)

    def remove(self):

        self.linkedlist.pop()

    def remove_front(self):

        self.linkedlist.popleft()

    def __getitem__(self, index):

        return self.linkedlist[index]

    def __setitem__(self, index, item):

        self.linkedlist[index] = item

    def __len__(self):

        return len(self.linkedlist)

def main():

    for linkedlist_class in [LinkedList1, LinkedList2]:

        linkedlist = linkedlist_class()

        print(len(linkedlist))

        linkedlist.add(1)
        linkedlist.add(2)
        linkedlist.add(3)

        print(linkedlist[2])

        linkedlist[1] = 4

        print(linkedlist[1])

        print(linkedlist[0])

        linkedlist.remove_front()

        linkedlist.add(5)

        linkedlist.remove()

        print(linkedlist[0])

        linkedlist.add_front(6)

        print(linkedlist[0])

        print(linkedlist[1])

        linkedlist.add(7)
        linkedlist.add(8)

        print(len(linkedlist))

        print("------")

if __name__ == '__main__':
    main()
